internal/gatewayapi/translator.go

```diff
@@ -273,28 +285,72 @@ func (t *Translator) Translate(resources *resource.Resources) (*TranslateResult,
 func (t *Translator) GetRelevantGateways(resources *resource.Resources) (
 	acceptedGateways, failedGateways []*GatewayContext,
 ) {
+	// make sure that the EnvoyProxyForGatewayClass is valid
+	if ep := resources.EnvoyProxyForGatewayClass; ep != nil {
+		err := validateEnvoyProxy(ep)
+		if err != nil {
+			t.Logger.Error(err, "Skipping GatewayClass because EnvoyProxy is invalid",
+				"gatewayclass", t.GatewayClassName,
+				"envoyproxy", ep.Name, "namespace", ep.Namespace)
+			status.SetGatewayClassAccepted(resources.GatewayClass,
+				false, string(gwapiv1.GatewayClassReasonInvalidParameters),
+				fmt.Sprintf("%s: %v", status.MsgGatewayClassInvalidParams, err))
+			return
+		}
+	}
+
 	for _, gateway := range resources.Gateways {
 		if gateway == nil {
 			panic("received nil gateway")
 		}
 
-		if gateway.Spec.GatewayClassName == t.GatewayClassName {
-			gc := &GatewayContext{
-				Gateway: gateway,
-			}
+		logKeysAndValues := []any{
+			"namespace", gateway.Namespace, "name", gateway.Name,
+		}
+		if gateway.Spec.GatewayClassName != t.GatewayClassName {
+			t.Logger.Info("Skipping Gateway because GatewayClassName doesn't match", logKeysAndValues...)
+			continue
+		}
+
+		gCtx := &GatewayContext{
+			Gateway: gateway,
+		}
 
-			// Gateways that are not accepted by the controller because they reference an invalid EnvoyProxy.
-			if status.GatewayNotAccepted(gc.Gateway) {
-				failedGateways = append(failedGateways, gc)
-			} else {
-				gc.ResetListeners(resources)
-				acceptedGateways = append(acceptedGateways, gc)
+		// Gateways that are not accepted by the controller because they reference an invalid EnvoyProxy.
+		if status.GatewayNotAccepted(gCtx.Gateway) {
+			failedGateways = append(failedGateways, gCtx)
+			t.Logger.Info("EnvoyProxy for Gateway not found ", logKeysAndValues...)
+			continue
+		}
+
+		gCtx.ResetListeners(resources)
+		if ep := gCtx.envoyProxy; ep != nil {
```

Comment:
so ` attachEnvoyProxy` lives inside `ResetListeners`, that sets `envoyProxy` to either EP for Gateway or EP for GatewayClass, this is going to validate twice, can we improve the way we validate, attach and update status ?